Handle both minimum and maximum duration for a turn

Cu3PO42 opened the pull request:
This is the PR promised in #117. It is an overhaul of `NanoTimer`, extending it to have both a minimum and maximum wait time. My understanding of the desired game logic is as follows:

- A turn takes AT MOST as long as permitted by the timeout.
- A turn takes AT LEAST as long as 1s/TPS to make sure the game does not run faster than the desired tick speed.
- A turn completes as soon as all bot intents are available.

The old implementation had no notion of two separate bounds. It simply set the turn timeout to `max(1/TPS, turnTimeout)`. The timer did not enforce a lower bound. However, a turn was only completed when the timer elapsed, making it both upper and lower bound at the same time.

While this does mean the game did not run faster than it should, it limits the maximal speed of the game. For the sake of this argument, let's assume we have configured max tick speed, meaning the configured turn timeout is active. Because we always needed to wait for the timer to elapse, the game could run no faster than 1/turnTimeout, even if the bot intents were all available much earlier.

The new implementation rectifies this and permits the game to run as fast as it can. It also makes the strategy for avoiding concurrent calls to onNextTurn more robust. As a general note, I am not convinced that the server in its entirety doesn't currently have concurrency issues unrelated to this PR. I'll need to do a more thorough investigation.

flemming-n-larsen commented on server/src/main/kotlin/dev/robocode/tankroyale/server/core/NanoTimer.kt:
```diff
@@ -1,58 +1,101 @@
+    override fun run() {
+        var lastTime = System.nanoTime()
+        if (minPeriodInNanos > 0) {
+            while (true) {
+                val now = System.nanoTime()
+                val diff = lastTime + minPeriodInNanos + totalPauseDuration.get() - now
+                val isPaused = pauseStartTime.get() != 0L
+                if (diff <= 0 && !isPaused) break
+                try {
+                    Thread.sleep(if (isPaused) Duration.ofSeconds(1) else Duration.ofNanos(diff))
+                } catch (e: InterruptedException) {
+                    if (jobExecuted.get()) return
+                }
+            }
+        }
+
+        while (true) {
+            if (jobExecuted.get()) break
+
             val now = System.nanoTime()
-            if (now - lastTime >= periodInNanos) {
-                lastTime = now
-                if (!isPaused.get()) {
-                    job.run()
+            val diff = lastTime + maxPeriodInNanos + totalPauseDuration.get() - now
+            val isPaused = pauseStartTime.get() != 0L
+            if (ready.get() || diff <= 0 && !isPaused) {
+                executeJob()
+                break
+            } else {
+                try {
+                    Thread.sleep(if (isPaused) Duration.ofSeconds(1) else Duration.ofNanos(diff))
+                } catch (e: InterruptedException) {
+                    continue
                 }
             }
         }
+
         thread = null
     }
```
I propose we split up the `run()` method into smaller methods to make it easier to read and to avoid redundant code as well following Clean Code principles. 🙂

```suggestion
    override fun run() {
        if (minPeriodInNanos <= 0) return

        waitForMinimumPeriod()
        waitAndExecuteJob()
        thread = null
    }

    private fun waitForMinimumPeriod() {
        val startTime = System.nanoTime()
        while (!canProceedAfterMinimumPeriod(startTime)) {
            sleepOrReturn()
        }
    }

    private fun waitAndExecuteJob() {
        val startTime = System.nanoTime()
        while (!jobExecuted.get()) {
            if (canExecuteJob(startTime)) {
                executeJob()
                break
            }
            sleep()
        }
    }

    private fun canProceedAfterMinimumPeriod(startTime: Long): Boolean {
        val now = System.nanoTime()
        val timeDiff = startTime + minPeriodInNanos + totalPauseDuration.get() - now
        return timeDiff <= 0 && !isPaused()
    }

    private fun canExecuteJob(startTime: Long): Boolean {
        val now = System.nanoTime()
        val timeDiff = startTime + maxPeriodInNanos + totalPauseDuration.get() - now
        return ready.get() || (timeDiff <= 0 && !isPaused())
    }

    private fun isPaused() = pauseStartTime.get() != 0L

    private fun sleepOrReturn() {
        try {
            if (isPaused()) {
                TimeUnit.SECONDS.sleep(1)
            } else {
                val sleepTime = calculateSleepTime(minPeriodInNanos)
                TimeUnit.NANOSECONDS.sleep(sleepTime)
            }
        } catch (e: InterruptedException) {
            if (jobExecuted.get()) return
        }
    }

    private fun sleep() {
        try {
            if (isPaused()) {
                TimeUnit.SECONDS.sleep(1)
            } else {
                val sleepTime = calculateSleepTime(maxPeriodInNanos)
                TimeUnit.NANOSECONDS.sleep(sleepTime)
            }
        } catch (ignore: InterruptedException) {
        }
    }

    private fun calculateSleepTime(periodInNanos: Long): Long {
        val now = System.nanoTime()
        val diff = periodInNanos + totalPauseDuration.get() - now
        return maxOf(diff, 0)
    }
````

Cu3PO42 replied:
I'm all in favor of making the main loop easier to read, but I believe your refactoring introduces a few minor issues.

* The check `if (minPeriodInNanos <= 0)` should only skip `waitForMinimumPeriod`, not the rest of the main method.
* `waitForMinimumPeriod` and `waitAndExecuteJob` need to share the same `startTime`. I suppose we could move it into a member.
* If we run into the `if (jobExecuted.get()) return` in `waitForMinimumPeriod` we should abort the entire execution, not just this method.
* We have double the number of `System.nanoTime()` calls. Since we're not doing a busy wait anymore, this is probably not a big deal. However, nanoTime is one of the most expensive timing APIs since it uses performance counters on x86. That said, it is the only simple monotonic clock available in Java, so I wouldn't replace it either.

I'll repair the first two problems and incorporate your changes.

flemming-n-larsen replied:
You are free to fix anything you fell for - especially, if I introduced bugs in my refactoring suggestion. 👍

Cu3PO42 commented:
I refactored the run method! I started with your changes and fixed the issues I pointed out above. While I was working on that, I spotted some other possible changes that I felt made sense. I know loops with empty bodies are sometimes controversial, but I like them better than putting the last condition as `if (foo) break` in the body. I'm open to other changes, though.

flemming-n-larsen commented:
> I refactored the run method! I started with your changes and fixed the issues I pointed out above. While I was working on that, I spotted some other possible changes that I felt made sense. I know loops with empty bodies are sometimes controversial, but I like them better than putting the last condition as `if (foo) break` in the body. I'm open to other changes, though.

I like all your changes. So I will merge it in now. 

I plan to do a release soon including this one. 😊👍
